controllers/flow_tree.go

```diff
@@ -0,0 +1,326 @@
+		if componentManager.allReadyOrUpdating() {
+			return stepResultMarkHappy
+		}
+		return stepResultMarkUnsatisfied
+	},
+}
+
+func buildFlowTree(components []ytv1.Component) *flowTree {
+	st := newSimpleStep
+	head := st(ytv1.UpdateStateNone)
+	tree := newFlowTree(head)
+
+	updMasterOrTablet := hasMaster(components) || hasTablets(components)
+	updMaster := hasMaster(components)
+	updTablet := hasMaster(components)
+	updStateless := hasStateless(components)
+
+	// TODO: if validation conditions can be not mentioned here or needed
+	tree.chain(
+		newConditionalForkStep(
+			ytv1.UpdateStatePossibilityCheck,
+			// This is the unhappy path.
+			st(ytv1.UpdateStateImpossibleToStart),
+			// Happy path will be chained automatically.
+		),
+	).chainIf(
+		updMasterOrTablet,
+		st(ytv1.UpdateStateWaitingForSafeModeEnabled),
+	).chainIf(
+		updTablet,
+		st(ytv1.UpdateStateWaitingForTabletCellsSaving),
+		st(ytv1.UpdateStateWaitingForTabletCellsRemovingStart),
+		st(ytv1.UpdateStateWaitingForTabletCellsRemoved),
+	).chainIf(
+		updMaster,
+		st(ytv1.UpdateStateWaitingForSnapshots),
+	).chain(
+		st(ytv1.UpdateStateWaitingForPodsRemoval),
+		st(ytv1.UpdateStateWaitingForPodsCreation),
+	).chainIf(
+		updMaster,
+		st(ytv1.UpdateStateWaitingForMasterExitReadOnly),
+	).chainIf(
+		updTablet,
+		st(ytv1.UpdateStateWaitingForTabletCellsRecovery),
+	).chainIfWithDeferredChain(
+		// TODO: we need to check for the exact component type here
+		updStateless,
+		// This is ugly, I suppose component code can be changed to avoid such transitions.
+		ytv1.UpdateStateWaitingForOpArchiveUpdate,
+		newConditionalForkStep(
+			ytv1.UpdateStateWaitingForOpArchiveUpdatingPrepare,
+			// This is the unhappy path.
+			// it will be chained with the next chained step via deferredChain argument
+			st(ytv1.UpdateStateWaitingForOpArchiveUpdate),
+			// Happy path will be chained automatically.
+		),
+	).chainIfWithDeferredChain(
+		// TODO: we need to check for the exact component type here
+		updStateless,
+		// This is ugly, I suppose component code can be changed to avoid such transitions.
+		ytv1.UpdateStateWaitingForQTStateUpdate,
+		newConditionalForkStep(
+			ytv1.UpdateStateWaitingForQTStateUpdatingPrepare,
+			// This is the unhappy path.
+			st(ytv1.UpdateStateWaitingForQTStateUpdate),
+			// Happy path will be chained automatically.
+		),
+	).chainIfWithDeferredChain(
+		// TODO: we need to check for the exact component type here
+		updStateless,
+		// This is ugly, I suppose component code can be changed to avoid such transitions.
+		ytv1.UpdateStateWaitingForYqlaUpdate,
+		newConditionalForkStep(
+			ytv1.UpdateStateWaitingForYqlaUpdatingPrepare,
+			// This is the unhappy path.
+			st(ytv1.UpdateStateWaitingForYqlaUpdate),
+			// Happy path will be chained automatically.
+		),
+	).chainIf(
+		updMaster,
+		st(ytv1.UpdateStateWaitingForSafeModeDisabled),
+	)
```

Comment:
Here we have full flow build with states in one function, so it is observable/rather easy to change.